effects/cargo/default.nix

```diff
@@ -0,0 +1,41 @@
+{ lib
+, mkEffect
+, cargo
+, cargoSetupHook
+}:
+
+# See docs/modules/ROOT/pages/reference/nix-functions/cargoPublish.adoc
+args@{ secretName ? throw ''effects.cargo: You must provide `secretName`, the name of the secret which holds the "${secretField}" field.''
+, secretField ? "token"
+, secretsMap ? { }
+, manifestPath ? "Cargo.toml"
+, src
+, extraPublishArgs ? [ ]
+, registryURL ? null
+, ...
+}:
+let
+  # This must match the "custom" identifier in ./cargo-setup-hook.sh
+  customRegistryIdentifier = "CUSTOM";
+in
+mkEffect (args // {
+  buildInputs = [ cargoSetupHook ];
+  inputs = [ cargo ];
+  secretsMap = { "cargo" = secretName; } // secretsMap;
+
+  env = args.env or { } // lib.optionalAttrs (registryURL != null) {
+    "CARGO_REGISTRIES_${customRegistryIdentifier}_INDEX" = registryURL;
+  };
+
+  # This style of variable passing allows overrideAttrs and modification in
+  # hooks like the userSetupScript.
```

Comment:
```suggestion
```

I don't think this was followed, but it doesn't seem necessary.
I think I'd prefer explicit functionality in this case.